### PR TITLE
feat: pipeline

### LINKED ISSUE #2257: Extended Query: Support Batch Execution
Hi!

The [Extended Query](https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY) protocol enables drivers to submit multiple BIND messages before SYNC. One of the big benefits of using Extended Queries is that you can efficiently send a batch of executions without waiting for round trips for each execution. Pairing that with prepared statements and some simplifications: you send a single PARSE, a bunch of BIND/EXECUTE and a SYNC to find out how things went.

In other words, you'd be able to support something like the following without needing 4 entire round trips. (I'm not _recommending_ this API since it would be a terrible breaking change.)

```javascript
await client.query({
  name: 'my_query',
  text: 'insert sometable (id, val) values ($1, $2)'
  values: [
    [ 1, 'asdf' ],
    [ 2, 'fdsa' ],
    [ 3, 'qwer' ],
    [ 4, 'uytr' ]
  ]
})
```

For more information, check out how the JDBC Postgres driver handles a batched execution. There are a few layers to dig through, but [this appears to be the core of the code that sends a batch of messages](https://github.com/pgjdbc/pgjdbc/blob/b80ebcef33f9c3801f01e28994746bc36abb2d21/pgjdbc/src/main/java/org/postgresql/core/v3/QueryExecutorImpl.java#L490-L509) and subsequently sends a single SYNC. NOTE: their driver imposes a limit of 128 records per batch as ([apparently](https://github.com/pgjdbc/pgjdbc/blob/b80ebcef33f9c3801f01e28994746bc36abb2d21/pgjdbc/src/main/java/org/postgresql/jdbc/PgPreparedStatement.java#L1604-L1608)) further batching does not improve performance.

### LINKED ISSUE #2646: implement pipelining mode allowed by libpq 14
libpq shipped with Postgres 14 now allows using pipelining mode 
https://www.postgresql.org/docs/14/libpq-pipeline-mode.html

as stated by these docs
> While the pipeline API was introduced in PostgreSQL 14, it is a client-side feature which doesn't require special server support and works on any server that supports the v3 extended query protocol.

So I don't know exactly when v3 extended query protocol has appeared (at least 2010 I will guess), but  it's there since a lot of time, so probably all supported version of pg server (9.6 and above) are able to support pipelining mode.

For the record, previous discussion on pipelining mode (before libpq supports it) is here :  https://github.com/brianc/node-postgres/issues/660

### PR DESCRIPTION
**Pull request by cesco69**
This PR implements pipeline mode, allowing multiple queries to be sent to the PostgreSQL server without waiting for the results of previous queries. This significantly reduces network latency, especially for batch operations or high-latency connections.

### Enabling Pipeline Mode

Pipeline mode is opt-in and can be enabled when creating a client or pool:

```js
// With Client
const client = new Client({ pipelineMode: true })
await client.connect()

// With Pool - all clients will have pipeline mode enabled
const pool = new Pool({ pipelineMode: true })
```

When pipeline mode is enabled, you can submit multiple queries and they will be sent to the server immediately:

```js
// All three queries are sent immediately to the server
const [result1, result2, result3] = await Promise.all([
  client.query('SELECT 1 as num'),
  client.query('SELECT 2 as num'),
  client.query('SELECT 3 as num'),
])
```

### Backpressure

When submitting a large number of queries in pipeline mode, you may want to limit how many queries are pending at once to prevent memory issues. The client supports configurable backpressure:

```js
const client = new Client({
  pipelineMode: true,
  pipelineMaxQueries: 100, // default is 1000
})
```

When the number of pending queries reaches `pipelineMaxQueries`, new queries are queued internally and will be sent once space becomes available. The client emits events to help you monitor backpressure:

```js
client.on('pipelineFull', () => {
  console.log('Pipeline is full, new queries will wait')
})

client.on('pipelineDrain', () => {
  console.log('Pipeline has drained, accepting new queries')
})
```

You can also check the current pipeline depth:

```js
console.log(client.pendingQueryCount) // number of queries waiting for results
```

The `pipelineDrain` event is emitted when the pending query count drops below 75% of `pipelineMaxQueries` (the "low water mark").

### Error Isolation

If one query fails in pipeline mode, other queries continue to execute:

```js
const results = await Promise.allSettled([
  client.query('SELECT 1 as num'),
  client.query('SELECT * FROM nonexistent_table'), // This will fail
  client.query('SELECT 3 as num'),
])

console.log(results[0].status) // 'fulfilled'
console.log(results[1].status) // 'rejected'
console.log(results[2].status) // 'fulfilled' - still succeeds!
```

### Prepared Statements in Pipeline Mode

Prepared statements work in pipeline mode, including concurrent queries with the same statement name:

```js
const results = await Promise.all([
  client.query({ name: 'get-user', text: 'SELECT $1::int as id', values: [1] }),
  client.query({ name: 'get-user', text: 'SELECT $1::int as id', values: [2] }),
  client.query({ name: 'get-user', text: 'SELECT $1::int as id', values: [3] }),
])
```

### Restrictions and Limitations

- **No COPY operations**: COPY commands are not supported in pipeline mode
- **No pg-cursor/pg-query-stream**: Custom submittables are not supported in pipeline mode
- **JavaScript client only**: Pipeline mode is not available in `pg-native`, would require:
  - https://github.com/brianc/node-libpq/pull/118 
  - https://github.com/brianc/node-postgres/pull/3596)

Note: Multi-statement queries (e.g., `SELECT 1; SELECT 2`) are rejected by PostgreSQL when using the extended query protocol, which pipeline mode uses. This is a PostgreSQL limitation.

### Transaction Behavior

Errors inside a transaction will abort subsequent queries until `ROLLBACK`. This is standard PostgreSQL behavior. If you send `COMMIT` on an aborted transaction, PostgreSQL automatically converts it to `ROLLBACK`:

```js
// In a transaction, if one query fails, subsequent queries also fail
await client.query('BEGIN')
await client.query('INSERT INTO t VALUES (1)')
await client.query('INSERT INTO t VALUES (1)') // fails: duplicate key
await client.query('INSERT INTO t VALUES (2)') // also fails: transaction aborted
await client.query('COMMIT') // returns ROLLBACK command
```

### When to Use Pipeline Mode

Pipeline mode is most beneficial when:
- You have many independent queries to execute
- Network latency is high (cloud databases, cross-region connections)
- You're doing batch operations

## Related Issues

- Closes https://github.com/brianc/node-postgres/issues/2646
- Closes https://github.com/brianc/node-postgres/issues/2257
- Closes https://github.com/brianc/node-postgres/pull/2942

## Benchmark

I have also updated the benchmark https://github.com/brianc/node-postgres/pull/3591/commits/b189ed517bbbc28f1e5c56bb3dfb3d05cdbc42e6

Query Type | Standard QPS | Pipeline QPS | Improvement
| -------- | ------- | ------- | ------- |
param | 374 | 440 | +18%
named | 380 | 440 | +16%
sequence | 1149 | 1680 | +46%
insert | 5623 | 11520 | +105%

### Raw Output
```
node bench.js
============================================================
STANDARD MODE BENCHMARK
============================================================
warmup done

param queries: 1868
qps: 374

named queries: 1899
qps: 380

sequence queries: 5745
qps: 1149

insert queries: 28115
qps: 5623


============================================================
PIPELINE MODE BENCHMARK
============================================================
warmup done

param queries: 2200
qps: 440

named queries: 2200
qps: 440

sequence queries: 8400
qps: 1680

insert queries: 57600
qps: 11520


============================================================
COMPARISON: Sequential vs Pipeline (same workload)
============================================================

Running 1000 sequential queries...
Sequential: 181ms (5538 qps)
Running 1000 pipeline queries...
Pipeline:   59ms (16842 qps)
```

### CODE REVIEW

**Comment by stephenh** [comment]
Ah wow, this is amazing!

@cesco69 just curious, in your example:

```
// All queries are sent immediately without waiting
const [r1, r2, r3] = await Promise.all([
  client.query('SELECT 1'),
  client.query('SELECT 2'),
  client.query('SELECT 3'),
])
```

Afaiu if `SELECT 2` breaks (invalid SQL query, an `UPDATE` that fails a constraint), then `SELECT 3` is aborted too.

...except that you have a test showing this is not the case (`nonexistent_table_xyz`). :thinking: 

I really thought the pg docs said that `SELECT 3 as num` would blow up, but I guess [reading again](https://www.postgresql.org/docs/current/libpq-pipeline-mode.html) they say "If any statement encounters an error, the server aborts the **current transaction** and does not execute any subsequent command in the queue until the next synchronization point".

So I guess if there is no open transaction, everything is fine. :+1: 

Speaking of transactions, I see you have a test for it, but does the `BEGIN` cause the current connection to be reserved to the current caller/request? It doesn't seem like it?

Like if I have two in-flight requests, I wouldn't want user1 issuing a `BEGIN` to then accept pipelined statements from `user2`.

Or I guess that's just on me to use a `Pool`, and only `BEGIN` on a reserved-to-that-user connection ... that makes sense/seems obvious in retrospect.

This is really great! :tada:

File: packages/pg/lib/client.js [hunk 1]
```diff
@@ -574,28 +652,109 @@ class Client extends EventEmitter {
...
-          })
+    if (this._pipelineMode) {
+      // In pipeline mode, we can send queries as soon as we're connected
+      // We don't need to wait for readyForQuery between queries
+      if (!this._connected) {
+        return
+      }
+      // In pipeline mode, send all queued queries immediately
+      while (this._queryQueue.length > 0) {
+        const query = this._queryQueue.shift()
```
Comment (charmander) [inline packages/pg/lib/client.js]: This should probably use a real queue with O(1) dequeue, even though the existing query queue doesn’t.

File: packages/pg/lib/query.js [hunk 2]
```diff
@@ -33,6 +34,11 @@ class Query extends EventEmitter {
   }
 
   requiresPreparation() {
+    // In pipeline mode, always use extended query protocol
```
Comment (charmander) [inline packages/pg/lib/query.js]: (So no client-side multi-statement check is needed at all?)

**Review by charmander — CHANGES_REQUESTED** [review]

File: packages/pg/lib/client.js [hunk 3]
```diff
@@ -666,6 +825,23 @@ class Client extends EventEmitter {
       query._result._types = this._types
     }
 
+    // Pipeline mode restriction: reject multi-statement queries
+    if (this._pipelineMode) {
+      const queryText = typeof config === 'string' ? config : config && config.text
+      if (queryText && typeof queryText === 'string' && queryText.includes(';')) {
```
[the remark this answers is not in the document]
Comment (cesco69) [reply packages/pg/lib/client.js]: fixed https://github.com/brianc/node-postgres/pull/3591/commits/0554631adfe09dfae73a311c948523c45ccf520c

File: packages/pg/lib/client.js [hunk 4]
```diff
@@ -451,19 +511,26 @@ class Client extends EventEmitter {
...
+    activeQuery._gotCommandComplete = true
     // delegate commandComplete to active query
     activeQuery.handleCommandComplete(msg, this.connection)
   }
 
   _handleParseComplete() {
-    const activeQuery = this._getActiveQuery()
+    const activeQuery = this._pipelineMode ? this._getCurrentPipelineQuery() : this._getActiveQuery()
     if (activeQuery == null) {
+      // In pipeline mode, parseComplete can arrive before we've processed the query
```
[the remark this answers is not in the document]
Comment (cesco69) [reply packages/pg/lib/client.js]: fix https://github.com/brianc/node-postgres/pull/3591/commits/c42122a19fc3e444927336a60bdee1232234a6d3

You're right, that check was wrong.

**Comment by cesco69** [comment]
> Ah wow, this is amazing!
> 
> @cesco69 just curious, in your example:
> 
> ```
> // All queries are sent immediately without waiting
> const [r1, r2, r3] = await Promise.all([
>   client.query('SELECT 1'),
>   client.query('SELECT 2'),
>   client.query('SELECT 3'),
> ])
> ```
> 
> Afaiu if `SELECT 2` breaks (invalid SQL query, an `UPDATE` that fails a constraint), then `SELECT 3` is aborted too.
> 
> ...except that you have a test showing this is not the case (`nonexistent_table_xyz`). 🤔
> 
> I really thought the pg docs said that `SELECT 3 as num` would blow up, but I guess [reading again](https://www.postgresql.org/docs/current/libpq-pipeline-mode.html) they say "If any statement encounters an error, the server aborts the **current transaction** and does not execute any subsequent command in the queue until the next synchronization point".
> 
> So I guess if there is no open transaction, everything is fine. 👍
> 
> Speaking of transactions, I see you have a test for it, but does the `BEGIN` cause the current connection to be reserved to the current caller/request? It doesn't seem like it?
> 
> Like if I have two in-flight requests, I wouldn't want user1 issuing a `BEGIN` to then accept pipelined statements from `user2`.
> 
> Or I guess that's just on me to use a `Pool`, and only `BEGIN` on a reserved-to-that-user connection ... that makes sense/seems obvious in retrospect.
> 
> This is really great! 🎉

Thanks! 🙏

You're right about error isolation, each query gets its own sync message, so errors don't cascade outside of transactions. The postgres docs refer to aborting "the current transaction", but in autocommit mode (no explicit BEGIN), each statement is its own transaction.

And yes, for transactions you'd use pool.connect() to get a dedicated client, same as without pipeline mode. Pipeline mode doesn't change connection ownership semantics, it just optimizes the wire protocol for that connection.

I have add more tests!

**Comment by cesco69** [comment]
# Benchmark

I have also updated the benchmark https://github.com/brianc/node-postgres/pull/3591/commits/b189ed517bbbc28f1e5c56bb3dfb3d05cdbc42e6


Query Type | Standard QPS | Pipeline QPS | Improvement
| -------- | ------- | ------- | ------- |
param | 374 | 440 | +18%
named | 380 | 440 | +16%
sequence | 1149 | 1680 | +46%
insert | 5623 | 11520 | +105%

## Raw Output
```
node bench.js
============================================================
STANDARD MODE BENCHMARK
============================================================
warmup done

param queries: 1868
qps: 374

named queries: 1899
qps: 380

sequence queries: 5745
qps: 1149

insert queries: 28115
qps: 5623


============================================================
PIPELINE MODE BENCHMARK
============================================================
warmup done

param queries: 2200
qps: 440

named queries: 2200
qps: 440

sequence queries: 8400
qps: 1680

insert queries: 57600
qps: 11520


============================================================
COMPARISON: Sequential vs Pipeline (same workload)
============================================================

Running 1000 sequential queries...
Sequential: 181ms (5538 qps)
Running 1000 pipeline queries...
Pipeline:   59ms (16842 qps)
```

**Comment by charmander** [comment]
You or @nigrosimone? And how much here is piped directly to and from AI? :/

**Comment by brianc** [comment]
I'll give this a proper & deep look tomorrow. One thing I did want to add now though before going through it, as it might inform your refactoring or touchups. I'm going to remove the queryQueue entirely pretty soon (like in a month or two) - one of the next things I plan on doing, early next week, is adding a deprecation warning if you queue a query while another query is in flight. In practice I can't see the query queue being used in "normal" code w/ async/await. I know in my own code I haven't used it in over a decade. It was a terrible decision from long ago when async patterns in node were still in their infancy.  So...please don't base any design decisions here (or really anywhere, ever 🙃 ) on the query queue implementation!

**Comment by nigrosimone** [comment]
> You or @nigrosimone?

Plot twist! We are the same user! :D

> And how much here is piped directly to and from AI? :/

I ran into issues I couldn’t solve on my own and got help from models like Opus 4.5 and ChatGPT 5. I’d say about 60% of the code is Ai.

In the past I have tried more time to implement pipeline:
- https://github.com/brianc/node-postgres/pull/3590
- https://github.com/brianc/node-postgres/pull/3534
- https://github.com/brianc/node-postgres/pull/3517

This time I decided to give it a try using AI (I’m not even a big fan myself), and I have to admit it was a huge help.

**Comment by charmander** [comment]
A twist indeed! I assume the PR description is also AI, but how about the latest comments and revisions? Some of them seemed conflicting and didn’t entirely make sense. It’s helpful to have context on this so I know how likely it is that given parts of the code exist for subtle reasons vs. being based on hallucinations, and whether my review suggestions are at risk of being applied with more bias than usual towards agreement when they might actually be wrong.

And yes, although this *could* go in before the query queue is removed (if it’s important that it makes it into pg 8.x…?), it might be a lot simpler without having the branching paths associated with the query queue.

**Comment by cesco69** [comment]
That’s a fair concern.

I want to clarify how I worked on this PR. I already contributed to pg in the past and I know the project reasonably well, but pipeline mode was something I never managed to implement before, because of the complexity and the many edge cases.

I did use AI, but not to just “generate the solution”. I started by writing a specification of what I wanted to achieve, listing the problematic cases that needed to be handled (error propagation, transactions, etc.), also including approaches I tried in the past and some ideas taken from other clients like Postgres.js and other old PR.

From there, I used AI to help me break this into a plan with many small tasks. Then I went through those tasks one by one and adding tests. The process was iterative and quite long, not a one-pass implementation. I was careful not to apply review suggestions automatically: I kept changes only when they made sense or test results. If some parts of the code look strange, it’s because I corrected earlier assumptions when edge cases became clearer, not because I blindly followed generated output.

About the query queue: I agree that pipeline mode would be simpler once the queue is removed. My goal here was to make it work with the current architecture.

If the goal in the future is to remove the queryQueue, this PR not make much sense, because it adds a lot of complexity and will make it harder to remove it later.
Also, without the queryQueue (and possibly without having to support callbacks, but only async/await promises), implementing pipeline mode would likely be much simpler.

Feel free to close the PR if you don't think it's valid!

**Comment by brianc** [comment]
> without having to support callbacks

I would love to drop callback support, but some users really cared about it a lot last time I suggested it, so probably gonna keep it around for now unless there's overwhelming support to remove them.

> Feel free to close the PR if you don't think it's valid!

Appreciate the PR and totally think it's valid! FWIW my own opinion on AI coding is I use it a _ton_ on personal projects. My approach is similar to your own. I don't have strong opinions on if it's used to help in contributions here provided the code looks good & in consistent format w/ the rest.  I will say massive github issues generated by AI aren't my favorite. I've seen an uptick in those recently..but its fine for now. I intend to set up so automations on issues here soon to help w/ some of that.

> My main concern here is should we wait for [9.0](https://github.com/brianc/node-postgres/discussions/3598) and destroy the queryQueue first? 

I don't want to invalidate your effort, though, work (and token cost, haha) - otoh waiting for a few weeks to month & having a much easier time of implementing it cleanly might be worth?  If that sounds good, I can go ahead & do a small patch now to deprecate the query queue on the 8.x branch, and we can cut a 9.x branch soon after to start work on the various breaking changes there.

**Comment by cesco69** [comment]
https://github.com/brianc/node-postgres/pull/3652
